---
title: "The Hoverfly microbiome: 16S data analysis"
author: "Pedro Rodrigues"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 3
    fig_caption: yes

fontsize: 14pt
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE)
```

```{r init, warning=FALSE, message=FALSE}
setwd("/Users/ento-user/Documents/Syrphids_microbiome/Syrphids_microbiome")
set.seed("12345678")
library(dada2);packageVersion("dada2") # 1.30.0
library(Biostrings); packageVersion("Biostrings") # 2.70.3
library(rmarkdown); packageVersion("rmarkdown") # 2.26
library(tidyverse); packageVersion("tidyverse") # 2.0.0
```

```{r more preparation, warning=FALSE, message=FALSE}
path.out <- "Figures/"
path.rds <- "RDS/"
theme_set(theme_bw())
```

Load Data (quality-trimmed and denoised using DADA2 and the nf-core/ampliseq pipeline)
```{r phyloseq-load-improved-data, eval=FALSE}
p_data.chimera_free <- readRDS("DADA2_table.rds")
```

```{r assign taxonomy, eval=FALSE}
# Assign taxonomy
tax <- assignTaxonomy(p_data.chimera_free, "./rdp_train_set_18.fa.gz", minBoot = 80, tryRC = TRUE, multithread=TRUE) # Slowest part
head(unname(tax))
tax <- addSpecies(tax, "./rdp_species_assignment_18.fa.gz")
```

```{r save-progress-taxonomy, eval=FALSE}
# Write to disk
saveRDS(tax, "RDS/tax_hoverflies.rds")
saveRDS(p_data.chimera_free, "RDS/DADA2_table.rds")
```

> NOTE: taxonomy will be improved later, using BLAST (nt database), but only after the ASV table is further cleaned in Phyloseq

## Move analysis over to Phyloseq

### Load RDS files (ASV table file and taxonomy file)
#### Also load metadata file, where sample names are row names
```{r phyloseq-load-data, eval=FALSE}
p_data.chimera_free <- readRDS("RDS/DADA2_table.rds")
tax <- readRDS("RDS/tax_hoverflies.rds")
metadata <- read.table("../input/2122metada.tsv", sep = "\t", header = TRUE)
samples.out <- data.frame(old_ID = rownames(p_data.chimera_free))
samples_out <- samples.out %>% separate_wider_delim(old_ID,delim ="_1.f", names = c("ID", "suffix"), cols_remove = FALSE)
metadata_merge <- left_join(metadata, samples_out, by = "ID")
row.names(metadata_merge) <- metadata_merge[,13]
#add isotope data to metadata
metadata_isotope <- read_csv("../input/metadata_isotope.csv")
metadata_merge_2 <- dplyr::left_join(metadata_merge, metadata_isotope, by="ID")
row.names(metadata_merge_2) <- metadata_merge_2[,13]
```

```{r backup-data, eval=FALSE}
write_tsv(metadata, "../input/metadataJan2025.tsv")
```

Make Phyloseq object
```{r phyloseq-make-object-1, eval=FALSE}
p_data <- phyloseq(otu_table(p_data.chimera_free, taxa_are_rows = FALSE),
                   sample_data(metadata_merge_2),
                   tax_table(tax))
```

Store DNA sequences in a different slot and simplify ASV names
```{r phyloseq-format-data, eval=FALSE}
dna <- Biostrings::DNAStringSet(taxa_names(p_data))
names(dna) <- taxa_names(p_data)
p_data <- merge_phyloseq(p_data, dna)
taxa_names(p_data) <- paste0("ASV", seq(ntaxa(p_data)))
p_data
```

Investigate taxa in the dataset - how many samples are not bacteria?
```{r summarize-no-bacteria, eval=FALSE}
p_data
p_data %>%
  subset_taxa(Kingdom!="Bacteria") %>%
  tax_table() %>% as.data.frame() %>% group_by(Class) %>% tally()
```

Not a lot of other taxa - 67 ASV identified as plants and two Archaea, from a total of 1056 ASVs

How many identified and unidentified ASV at the Phylum and Class levels?

```{r summarize-no-bacteria-phylum-class, eval=FALSE}
p_data
p_data %>%
  tax_table() %>% as.data.frame() %>% group_by(Phylum, Class) %>% tally() %>% paged_table()
```

A few NA. Also, some chloroplasts (and possibly mitochondria, but not possible to tell yet). Let's exclude these obvious contaminants before saving sequences as a fasta file and using Blast to (hopefully) get higher definition for taxa currently classified as "NA".

First, I will make data easier to work with, by transforming the phyloseq object into a summary table.

```{r relative-abundance-dada2-processed-data-and-taxonomy, eval=FALSE}
pdata_df <- p_data %>% psmelt() %>% arrange(OTU) %>% dplyr::rename(ASV = OTU) %>% spread(Sample, Abundance)
```

Let's remove chloroplasts, plant DNA and other taxa that may not be relevant here.

```{r remove-chloroplast--plants-archaea, eval=FALSE}
p_data_taxlean <- p_data %>% subset_taxa(Kingdom != "Archaea") %>% subset_taxa((Order != "Chloroplast") | is.na(Order)) %>% subset_taxa((Kingdom != "Eukaryota") | is.na(Kingdom)) %>% subset_taxa((Class != "Cyanobacteria") | is.na(Order)) %>% subset_samples(sample_sums(p_data) > 0)
```

Save taxonomy table and ASV fasta and blast unknown sequences to increase accuracy of classification

```{r save-taxonomy-table, eval=FALSE}
p_data_taxlean %>%
  tax_table() %>%
  write.csv("./Syrphid_asv_02_2025.csv")
```

```{r save-sequences-fasta, eval=FALSE}
p_data_taxlean %>%
      refseq() %>%
      Biostrings::writeXStringSet("./Syrphid_asv_02_2025.fasta", append=FALSE,
                                  compress=FALSE, compression_level=NA, format="fasta")
```
