---
title: "The Hoverfly microbiome: 16S data analysis"
author: "Pedro Rodrigues"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 3
    fig_caption: yes

fontsize: 14pt
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE)
```

```{r init, warning=FALSE, message=FALSE}
setwd("/Users/ento-user/Documents/Syrphids_microbiome/Syrphid_microbiome")
set.seed("12345678")
library(dada2);packageVersion("dada2") # 1.30.0
library(Biostrings); packageVersion("Biostrings") # 2.70.3
library(rmarkdown); packageVersion("rmarkdown") # 2.26
library(tidyverse); packageVersion("tidyverse") # 2.0.0
```

```{r more preparation, warning=FALSE, message=FALSE}
path.out <- "Figures/"
path.rds <- "RDS/"
theme_set(theme_bw())
```

Load Data (quality-trimmed and denoised using DADA2 and the nf-core/ampliseq pipeline)

```{r phyloseq-load-improved-data, eval=FALSE}
p_data.chimera_free <- readRDS("DADA2_table.rds")
```

```{r assign taxonomy, eval=FALSE}
# Assign taxonomy
tax <- assignTaxonomy(p_data.chimera_free, "./rdp_train_set_18.fa.gz", minBoot = 80, tryRC = TRUE, multithread=TRUE) # Slowest part
head(unname(tax))
tax <- addSpecies(tax, "./rdp_species_assignment_18.fa.gz")
```

```{r save-progress-taxonomy, eval=FALSE}
# Write to disk
saveRDS(tax, "RDS/tax_hoverflies.rds")
saveRDS(p_data.chimera_free, "RDS/DADA2_table.rds")
```

> NOTE: taxonomy will be improved later, using BLAST (nt database), but only after the ASV table is further cleaned in Phyloseq

## Move analysis over to Phyloseq

### Load RDS files (ASV table file and taxonomy file)

#### Also load metadata file, where sample names are row names

```{r phyloseq-load-data, eval=FALSE}
p_data.chimera_free <- readRDS("RDS/DADA2_table.rds")
tax <- readRDS("RDS/tax_hoverflies.rds")
metadata <- read.table("../input/2122metada.tsv", sep = "\t", header = TRUE)
samples.out <- data.frame(old_ID = rownames(p_data.chimera_free))
samples_out <- samples.out %>% separate_wider_delim(old_ID,delim ="_1.f", names = c("ID", "suffix"), cols_remove = FALSE)
metadata_merge <- left_join(metadata, samples_out, by = "ID")
row.names(metadata_merge) <- metadata_merge[,13]
#add isotope data to metadata
metadata_isotope <- read_csv("../input/metadata_isotope.csv")
metadata_merge_2 <- dplyr::left_join(metadata_merge, metadata_isotope, by="ID")
row.names(metadata_merge_2) <- metadata_merge_2[,13]
```

```{r backup-data, eval=FALSE}
write_tsv(metadata, "../input/metadataJan2025.tsv")
```

Make Phyloseq object

```{r phyloseq-make-object-1, eval=FALSE}
p_data <- phyloseq(otu_table(p_data.chimera_free, taxa_are_rows = FALSE),
                   sample_data(metadata_merge_2),
                   tax_table(tax))
```

Store DNA sequences in a different slot and simplify ASV names

```{r phyloseq-format-data, eval=FALSE}
dna <- Biostrings::DNAStringSet(taxa_names(p_data))
names(dna) <- taxa_names(p_data)
p_data <- merge_phyloseq(p_data, dna)
taxa_names(p_data) <- paste0("ASV", seq(ntaxa(p_data)))
p_data
```

Investigate taxa in the dataset - how many samples are not bacteria?

```{r summarize-no-bacteria, eval=FALSE}
p_data
p_data %>%
  subset_taxa(Kingdom!="Bacteria") %>%
  tax_table() %>% as.data.frame() %>% group_by(Class) %>% tally()
```

Not a lot of other taxa - 67 ASV identified as plants and two Archaea, from a total of 1056 ASVs

How many identified and unidentified ASV at the Phylum and Class levels?

```{r summarize-no-bacteria-phylum-class, eval=FALSE}
p_data
p_data %>%
  tax_table() %>% as.data.frame() %>% group_by(Phylum, Class) %>% tally() %>% paged_table()
```

A few NA. Also, some chloroplasts (and possibly mitochondria, but not possible to tell yet). Let's exclude these obvious contaminants before saving sequences as a fasta file and using Blast to (hopefully) get higher definition for taxa currently classified as "NA".

First, I will make data easier to work with, by transforming the phyloseq object into a summary table.

```{r relative-abundance-dada2-processed-data-and-taxonomy, eval=FALSE}
pdata_df <- p_data %>% psmelt() %>% arrange(OTU) %>% dplyr::rename(ASV = OTU) %>% spread(Sample, Abundance)
```

Let's remove chloroplasts, plant DNA and other taxa that may not be relevant here.

```{r remove-chloroplast--plants-archaea, eval=FALSE}
p_data_taxlean <- p_data %>% subset_taxa(Kingdom != "Archaea") %>% subset_taxa((Order != "Chloroplast") | is.na(Order)) %>% subset_taxa((Kingdom != "Eukaryota") | is.na(Kingdom)) %>% subset_taxa((Class != "Cyanobacteria") | is.na(Order)) %>% subset_samples(sample_sums(p_data) > 0)
```

Save taxonomy table and ASV fasta and blast unknown sequences to increase accuracy of classification

```{r save-taxonomy-table, eval=FALSE}
p_data_taxlean %>%
  tax_table() %>%
  write.csv("./Syrphid_asv_02_2025.csv")
```

```{r save-sequences-fasta, eval=FALSE}
p_data_taxlean %>%
      refseq() %>%
      Biostrings::writeXStringSet("./Syrphid_asv_02_2025.fasta", append=FALSE,
                                  compress=FALSE, compression_level=NA, format="fasta")
```

Blast was run using the script "blast_ASV_syrphid.sh", and taxonomixal ranks were added to the output file ("0203205_nt_blast_syrphid.tsv") using the script "Get_lineage_blast.sh". The resulting spreadsheet was saved as "syr16S_annotated_ASV_nt_blast.csv". Next, I will identify "NA" sequences that were resolved by blast, and manually inspect their identification and eliminate any spurious results (e.g. annotations containing mistakes, non-taxonomical attributions, etc).

```{r load taxonomical data}
old_tax_table <- read.csv("../results/blast/Syrphid_asv_02_2025.csv")
blast_tax_table <- read.csv("../results/blast/syr16S_annotated_ASV_nt_blast.csv")
NA_old_tax_table <- old_tax_table %>% filter(is.na(Genus)) %>% 
  rename("X" = "qseqid") #select only ASV with NA at the Genus level; rename ASV id column to match the corresponding blast column name
NoNA_blast_tax_table <- blast_tax_table %>% filter(!is.na(genus)) # select blast results that resolved identification at the Genus level
resolved_by_blast <- left_join(NA_old_tax_table, NoNA_blast_tax_table, by = "qseqid") %>% filter(!is.na(genus)) # combine both subset data sets, using the old tax table as the reference for rows to include, and qseqid as the column in common to allow the merge of these tables; next, exclude "genus" in blast results that were not resolved
```

Blast helped resolve the identification of 164 ASV, including bacteria and contaminants (e.g. plants, fungus). Next, I will transfer all taxonomical levels for resolved genera from blast results to the "old_tax_table".

```{r update taxonomy}
# Prepare resolved_by_blast to be added to the final taxonomy file
# Select only relevant columns
resolved_by_blast_slim <- resolved_by_blast %>% select(c("qseqid", "superkingdom", "phylum", "class", "order", "family", "genus", "species"))
# Rename columns to match phyloseq tax table
colnames(resolved_by_blast_slim) <- c("X", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
# Rename values in "Species" to eliminate characters that Phyloseq might not like
resolved_by_blast_slim <- resolved_by_blast_slim %>% mutate(Species = paste ("sp"))
# Use anti-join to eliminate from old_tax_table entries that overlap with the new results
old_tax_table_slice <- anti_join(old_tax_table, resolved_by_blast_slim, by = "X")
# Now join rows from both datasets into the final dataset
updated_tax <- bind_rows(resolved_by_blast_slim, old_tax_table_slice)
```

For now this should be sufficient. Downstream, I will further improve taxonomy by replacing NA at the genus level with the next best taxonomical id followed by "sp" (e.g. "Orbacaeae sp.", or "Bacteria sp.", etc).

Add new taxonomy back to the phyloseq object

```{r add-updated-taxonomy}
#rename rows with ASV identifier
row.names(updated_tax) <- updated_tax$X
# remove first column
updated_tax <- updated_tax[-1]
new_tax <- tax_table(as.matrix(updated_tax))
tax_table(p_data_taxlean) <- new_tax
```

Further cleaning - now let's eliminate resolved IDs that revealed more contaminants in the dataset

```{r what-taxa-is-there-2, eval=FALSE}
updated_tax %>% select(Kingdom) %>% unique()
updated_tax %>% select(Phylum) %>% unique()
updated_tax %>% select(Class) %>% unique()
updated_tax %>% select(Order) %>% unique()
```

```{r further-cleaning}
p_data_taxlean <- p_data_taxlean %>% subset_taxa(Kingdom != "Eukaryota" | is.na(Kingdom)) %>% subset_taxa(Phylum != "Cyanobacteria/Chloroplast" | is.na(Phylum)) %>% subset_taxa(Class != "Cyanobacteria" | is.na(Class))
```

Let's check if there is still contaminants left

```{r double-check-contaminants, eval=FALSE}
p_data_taxlean %>% psmelt() %>% select(Genus) %>% unique()
p_data_taxlean %>% psmelt() %>% select(Class) %>% unique()
p_data_taxlean %>% psmelt() %>% select(Phylum) %>% unique()
p_data_taxlean %>% psmelt() %>% select(c("OTU","Phylum")) %>% filter(is.na(Phylum)) %>% select(OTU) %>% unique()
```

No obvious contaminant, although taxa with "NA" at the phylum level are suspicious (n=47). However, ASV2 was at first "NA" at the phylum level and after Blast identified as Bacteria, so we can not rule out (yet) that these ASVs are not contaminants. In downstream analysis some of those may be excluded due to low abundance in the dataset.

Before proceeding to stats and plots, save the new taxonomy table and a summary of the data

```{r save-updated_taxonomy, eval=FALSE}
saveRDS(p_data_taxlean,"../results/RDS/p_data_taxlean.rds")
```

```{r save-ASV-metadata-seq-table, eval=FALSE}
write.table(p_data_taxlean %>% psmelt() %>% arrange(OTU) %>% 
              dplyr::rename(ASV = OTU) %>% select(ASV, Kingdom, Phylum, Class, Order, Family, Genus, Sample, Abundance) %>% spread (Sample, Abundance), 
    file = "../results/summary_syrphid16S_February2025.tsv", sep = "\t", quote = F, row.names = F, col.names = T)
```


----------------
#Stats and Plots
----------------

```{r load-more-packages}
library(microbiome)
library(ggpubr)
library(knitr)
library(ggtext)
library(scales)
library(ggside)
library(RColorBrewer)
#devtools::install_github("vmikk/metagMisc")
library(metagMisc)
```

First, check the abundance of Wolbachia in the samples.

```{r Wolbachia prevalence across samples}
# Join the pseq_barplot and metadata_merge
Wolbachia_barplot <- tax_glom(p_data_taxlean, "Genus") %>% psmelt() %>% 
  mutate(BacSpec = ifelse(!Genus == "Wolbachia", "Other", "Wolbachia"))

# Create the side-by-side bar plots using ggplot2 and facet_wrap
ggplot(Wolbachia_barplot, aes(x = Sample, y = Abundance, fill = BacSpec)) +
  geom_bar(stat = "identity", position = "fill", width = 0.9) +
  facet_wrap(~sample_Species, scale="free") + 
  labs(title = "Hoverfly microbiome is dominated by Wolbachia",
       x = "Samples",
       y = "Relative Frequency") +
  theme(legend.position = "top", 
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank()) +
  labs(fill = "Bacteria Species") +
  scale_fill_manual(values = c("Wolbachia" = "#009E73", "Other" = "darkgrey"))
ggsave("../results/Figures/WolbachiaPrevalence.pdf", width=15, height=10)
```

How many samples are left after removing Wolbachia?

```{r leftover-after-removing-Wolbachia}
leftover_barplot <- tax_glom(p_data_taxlean, "Genus") %>% subset_taxa(Genus != "Wolbachia") %>% psmelt()
leftover_barplot %>% filter(sample_Species == "Allograpta obliqua") %>% group_by(Sample) %>% summarise(sum_abund = sum(Abundance)) %>% arrange(sum_abund)
leftover_barplot %>% filter(sample_Species == "Eupeodes americanus") %>% group_by(Sample) %>% summarise(sum_abund = sum(Abundance)) %>% arrange(sum_abund)
```

How many samples would be lost if we keep only leftover samples with a size greater than 1000 sequences?
```{r samples-lost-after-AbundanceFilter}
leftover_barplot %>% filter(sample_Species == "Allograpta obliqua") %>% group_by(Sample) %>% summarise(sum_abund = sum(Abundance)) %>%
  filter(sum_abund > 1000) %>% nrow()
leftover_barplot %>% filter(sample_Species == "Eupeodes americanus") %>% group_by(Sample) %>% summarise(sum_abund = sum(Abundance)) %>%
  filter(sum_abund > 1000) %>% nrow()
```

With this arbitrary filter (1000 sequences), there is still a decent number of samples left in the dataset for each host species. The main concern then is whether these samples are still representative of the population, i.e. is the microbiota of samples with lower counts of Wolbachia different from samples with higher counts of Wolbachia, or are the titers of Wolbachia independent of the gut microbiota composition? With this current dataset we are not able to make this distinction, but we can infer that if we do find differences supporting our hypotheses regarding the relationship between host diet and their migratory behavior, then we can at the very least infer that these differences are true, for this subgroup of samples.

What is sample size distribution for each species after removing Wolbachia and filtering abundance > 1000 ?
```{r plot-sample-size-noWolbachia}  
# Create the side-by-side bar plots using ggplot2 and facet_wrap
ggplot(leftover_barplot %>% group_by(Sample, sample_Species) %>% summarize(sum_abund = sum(Abundance), .groups = "drop") %>%
         filter(sum_abund > 1000) %>% arrange(sample_Species, desc(sum_abund)) %>%
         unite("sp_sample", sample_Species, Sample, sep = "_", remove=FALSE) %>%
         data.frame() %>%
         mutate(sp_sample = factor(sp_sample, levels = sp_sample)) %>% group_by(sample_Species), aes(x = sp_sample, y = sum_abund)) +
  geom_col() +
  geom_hline(yintercept = 1000, color = "white") +
  facet_wrap(~sample_Species, scale="free") + 
  labs(title = "Sample size distribution after removal of Wolbachia",
       x = "Samples",
       y = "Sum of read counts") +
  theme(legend.position = "top", 
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank())
ggsave("../results/Figures/NoWol_samplesizedistrib.pdf", width=15, height=10)
```

Most samples have over 1000 sequences, with close to half having over 4-5K sequences each. I think this is a good sign, supporting the decision of eliminating Wolbachia from the dataset, eliminating possible noise/masking of biologically-relevant differences in microbiome composition, and allowing for additional exploration of the Syrphid microbiome.

> A note on further filtering data: I decided not to filter data based on prevalence, after noticing that the number of OTU would dramatically decrease, even when using "soft" filtering criteria (e.g. "exclude ASV with total abundance smaller than 10 and prevalence in less than 2.5% of the samples"). The reason the loss was so dramatic is possibly because many ASV are the same bacteria (Genus), and ASV does tend to overestimate richness due to variation in 16S nucleotide composition within species that DADA2 classifies as different ASV, so generating many ASVs for the same Genus of bacteria, with a small number of reads for most ASV variants of the same Genus, and a larger count of reads for the Genus ASV variant that is most common. Therefore, perhaps for hypothesis testing or plotting, it may make more sense to agglomerate data to Genus and filter, but for other types of analysis perhaps variation in the 16S, even within a Genus of bacteria, might be biologically relevant. For instance, this kind of variation may be informative if variation is associated to location where samples were collected.

<!-- 1. Prevalence filtering: I will eliminate any ASV with that in average less than "10" across at least 5% of the samples -->

<!-- ```{r eliminate low prevalence ASV} -->
<!-- filtered_and_clean <- phyloseq_filter_prevalence(p_data_taxlean, prev.trh = 0.025, abund.trh = 10, threshold_condition = "OR") #p_data_tax_lean has 807 taxa (ASV) -->
<!-- ``` -->

<!-- After this filter, 459 taxa were kept in the dataset. This is a loss of approximately  -->

So, based on these results, I will split data into two datasets: with and without Wolbachia.

```{r split-data-byWolb}
Wolb <- p_data_taxlean
NoWolb <- p_data_taxlean %>% subset_taxa(Genus != "Wolbachia") 
NoWolb <- NoWolb %>% subset_samples(sample_sums(NoWolb) > 1000)
```

For rarefying data, I will use phyloseq_mult_rare_avg, that will subsample samples 1000 times and average the number of times ASVs are observed in each sample according to the threshold I will establish below

```{r thresholds-rarefy-data, eval=FALSE}
#Wolb
otu_table(Wolb) <- otu_table(t(otu_table(p_data_taxlean)), taxa_are_rows=TRUE)
sample_size_quantile <- as.data.frame(quantile(sample_sums(Wolb), probs = c(0, 0.10, 0.15, 0.225, 0.25, 0.5, 0.75, 1)))
sample_size_quantile["10%", ] # 3,656 reads
sample_size_quantile["25%", ] # 14,021 reads
sample_size_quantile["50%", ] # 25,268.5 reads
sample_size_quantile["75%", ] # 32,769 reads
#NoWolb
NoWolb_1 <- NoWolb
otu_table(NoWolb_1) <- otu_table(t(otu_table(NoWolb)), taxa_are_rows=TRUE)
sample_size_quantile <- as.data.frame(quantile(sample_sums(NoWolb_1), probs = c(0, 0.10, 0.15, 0.225, 0.25, 0.5, 0.75, 1)))
sample_size_quantile["10%", ] # 1,395.8 reads
sample_size_quantile["25%", ] # 2,238 reads
sample_size_quantile["50%", ] # 5,492 reads
sample_size_quantile["75%", ] # 9,589.5 reads
```

Samples that would be lost with some of these thresholds

```{r samples-lost}
Wolb_loss <- as.data.frame(sample_sums(Wolb))
names(Wolb_loss) <- "Sample_size"
NoWolb_loss <- as.data.frame(sample_sums(NoWolb))
names(NoWolb_loss) <- "Sample_size"
data.frame(Wolbachia = c("yes", "yes", "yes", "no", "no", "no"), Quantile_cutoff = c("no cutoff", "10%", "25%", "no cutoff", "10%", "25%"), Samples_left = c(nrow(Wolb_loss), nrow(filter(Wolb_loss, Sample_size > 3656)), nrow(filter(Wolb_loss, Sample_size > 14021)), nrow(NoWolb_loss), nrow(filter(NoWolb_loss, Sample_size > 2238)), nrow(filter(NoWolb_loss, Sample_size > 5492))))
```

With Wolbachia, sample size is still robust even when using a quantile cutoff of 25% (that is, keeping samples representing the top 75% of sample size range). But the "no Wolbachia" dataset is much smaller and even a small cutoff (10% quantile) can take away important samples that perhaps would be informative for our questions related to sample geography, migration and diet. Therefore, we should try to keep as many samples as possible in the "no Wolbachia" dataset. Perhaps using the smallest samples as the threshold for rarefaction is best.

```{r rarefy-samples}
Wolb_rr <- phyloseq_mult_raref_avg(p_data_taxlean, SampSize = 14021, iter=1000, parallel = TRUE,verbose = TRUE)
NoWolb_rr <- phyloseq_mult_raref_avg(NoWolb, SampSize = sample_size_quantile["0%", ], iter=1000, parallel = TRUE,verbose = TRUE)
Wolb_rr <- transform_sample_counts(Wolb_rr, function(x){ x * 14021})
NoWolb_rr <- transform_sample_counts(NoWolb_rr, function(x){x * sample_size_quantile["0%", ]})
```

One interesting (but expected) result from rarefying data: keeping Wolbachia, and more samples, but rarefying at a higher threshold resulted in fewer taxa (381) than removing Wolbachia, but using a smaller rarefaction cutoff for samples size (1021 sequences, 401 taxa kept).

The phyloseq_mult_raref_avg function removes the "Refseq" component of the phyloseq object. I will put it back to each object (Wolb and NoWolb), so that it can be used in downstream analysis

```{r add-back-sequences}
Wotu_table <- t(as(otu_table(Wolb_rr), "matrix"))
asv_ids <- colnames(Wotu_table)
refseq_phyloseq <- refseq(p_data_taxlean)
refseq_phyloseq_W <- refseq_phyloseq[names(refseq_phyloseq) %in% asv_ids]
Wolb_rr_up <- merge_phyloseq(Wolb_rr, refseq_phyloseq_W)
NoWotu_table <- t(as(otu_table(NoWolb_rr), "matrix"))
asv_ids <- colnames(NoWotu_table)
refseq_phyloseq_NoW <- refseq_phyloseq[names(refseq_phyloseq) %in% asv_ids]
NoWolb_rr_up <- merge_phyloseq(NoWolb_rr, refseq_phyloseq_NoW)
```


Next, let's test two hypothesis:
(1) That microbial communities are different between _Allograpta_ and _Eupeodes_ _americanus_
(2) That microbial communities are different between North and South populations for *non-migratory* hoverflies ( _Allograpta_ ), but more similar between North and South populations for *migratory* hoverflies ( _Eupeodes_ ) 

To test these hypotheses, I will use a Permanova test. One assumption of Permanova is that there is homogeneity of dispersion between groups.
For hypothesis "1", host species is the grouping variable, and for the second hypothesis, for each species, region of sampling (north vs south) is the grouping variable.


```{r betadisper}
library("vegan")
# Extract OTU table from phyloseq object, calculate Bray-Curtis distances, and calculate beta dispersion
# 1. For Wolb_rr_up; first, exclude Eupeodes pomus from the dataset (only one species)
Wolb_rr_up_filt <- Wolb_rr_up %>% subset_samples(Species != "Eupeodes pomus")
otu_table <- t(as(otu_table(Wolb_rr_up_filt), "matrix"))
Wolb_bray_dist <- vegdist(otu_table, method = "bray")
metadata_Wolb_rr_up_filt <- data.frame(sample_data(Wolb_rr_up_filt))
disper_SpID <- betadisper(Wolb_bray_dist, metadata_Wolb_rr_up_filt$Species)

# Test if dispersion between groups is significant
anova(disper_SpID) # P-value VERY significant (P<0.00000001); data is overdispersed

# 2. For NoWolb_rr_up
NoWolb_rr_up_filt <- NoWolb_rr_up %>% subset_samples(Species != "Eupeodes pomus")
otu_table <- t(as(otu_table(NoWolb_rr_up_filt), "matrix"))
NoWolb_bray_dist <- vegdist(otu_table, method = "bray")
metadata_NoWolb_rr_up_filt <- data.frame(sample_data(NoWolb_rr_up_filt))
disper_SpID <- betadisper(NoWolb_bray_dist, metadata_NoWolb_rr_up_filt$Species)

# Test if dispersion between groups is significant
anova(disper_SpID) # p > 0.05, data is not overdispersed, so we can use bray distances for the data with Wolbachia removed
```

Homogeneity of dispersion is not met for the Wolbachia-present dataset. I will try to use Unifrac distances, but first I need to build a phylogenetic tree. I will also calculate Unifrac distances for the Wolbachia-absent dataset, in case needed for downstream analyses.

1. Build tree and calculate unifrac distances
```{r UniFrac-distances}
library(phangorn)
library(DECIPHER)
library(GUniFrac)
# Wolb_rr_up
refseq_phyloseq_W_filt <- refseq(Wolb_rr_up_filt)
alignment <- AlignSeqs(refseq_phyloseq_W_filt, anchor=NA)
phangAlign <- phyDat(as(alignment, "matrix"), type="DNA")
dm <- dist.ml(phangAlign)
treeNJ <- NJ(dm)
mod.test <- modelTest(phangAlign) 
mod.test$Model[mod.test$AIC==min(mod.test$AIC)] #this gives GTR+G(4)+I as the best model, i.e. the model with the smallest AIC (or BIC) result
fit = pml(treeNJ, data=phangAlign)
fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
        rearrangement = "stochastic", control = pml.control(trace = 0)) #this may take 15-20 min
rooted_fitGTR_tree <- midpoint(fitGTR$tree)
#detach("package:phangorn", unload=TRUE)
Wolb_rr_up_phytree <- merge_phyloseq(Wolb_rr_up_filt, rooted_fitGTR_tree)


# NoWolb_rr_up
refseq_phyloseq_NoW_filt <- refseq(NoWolb_rr_up_filt)
alignment <- AlignSeqs(refseq_phyloseq_NoW_filt, anchor=NA)
phangAlign <- phyDat(as(alignment, "matrix"), type="DNA")
dm <- dist.ml(phangAlign)
treeNJ <- NJ(dm)
mod.test <- modelTest(phangAlign) 
mod.test$Model[mod.test$AIC==min(mod.test$AIC)] #this gives GTR+G(4)+I as the best model, i.e. the model with the smallest AIC (or BIC) result
fit = pml(treeNJ, data=phangAlign)
fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
        rearrangement = "stochastic", control = pml.control(trace = 0)) #this may take 15-20 min
rooted_fitGTR_tree <- midpoint(fitGTR$tree)
#detach("package:phangorn", unload=TRUE)
NoWolb_rr_up_phytree <- merge_phyloseq(NoWolb_rr_up_filt, rooted_fitGTR_tree)
```

2. Next, I calculate if data is still overdispersed using betadisper
```{r betadisper-UniFrac}
#unifrac_dist <- UniFrac(Wolb_rr_up_phytree, weighted = TRUE)
#Wolb
metadata <- data.frame(sample_data(Wolb_rr_up_phytree))
unifracs = GUniFrac::GUniFrac(t(as(otu_table(Wolb_rr_up_phytree), "matrix")), phy_tree(Wolb_rr_up_phytree), alpha = c(0, 0.5, 1))$unifracs
Wolb_unifrac_distUW <- as.dist(unifracs[, , "d_UW"])
Wolb_unifrac_dist_d_1 <- as.dist(unifracs[, , "d_1"])
Wolb_unifrac_dist_d_05 <- as.dist(unifracs[, , "d_0.5"])
Wolb_unifrac_betadisperUW <- betadisper(Wolb_unifrac_distUW, metadata$Species) 
anova(Wolb_unifrac_betadisperUW)
Wolb_unifrac_betadisper_d_1 <- betadisper(Wolb_unifrac_dist_d_1, metadata$Species) 
anova(Wolb_unifrac_betadisper_d_1)
Wolb_unifrac_betadisper_d_05 <- betadisper(Wolb_unifrac_dist_d_05, metadata$Species) 
anova(Wolb_unifrac_betadisper_d_05)

metadata <- data.frame(sample_data(NoWolb_rr_up_phytree))
unifracs = GUniFrac::GUniFrac(t(as(otu_table(NoWolb_rr_up_phytree), "matrix")), phy_tree(NoWolb_rr_up_phytree), alpha = c(0, 0.5, 1))$unifracs
NoWolb_unifrac_distUW <- as.dist(unifracs[, , "d_UW"])
NoWolb_unifrac_dist_d_1 <- as.dist(unifracs[, , "d_1"])
NoWolb_unifrac_dist_d_05 <- as.dist(unifracs[, , "d_0.5"])
NoWolb_unifrac_betadisperUW <- betadisper(NoWolb_unifrac_distUW, metadata$Species) 
anova(NoWolb_unifrac_betadisperUW)
NoWolb_unifrac_betadisper_d_1 <- betadisper(NoWolb_unifrac_dist_d_1, metadata$Species) 
anova(NoWolb_unifrac_betadisper_d_1)
NoWolb_unifrac_betadisper_d_05 <- betadisper(NoWolb_unifrac_dist_d_05, metadata$Species) 
anova(NoWolb_unifrac_betadisper_d_05)

```

Unweighted Unifrac distances worked best for both datasets, with or without Wolbachia. Because unweighted distances resolved homogeneity of variance across datasets, and because Bray-Curtis resolved only marginally (p=0.06) homogeneity of variance for groups in the "No Wolbachia" samples, I will adopt unifrac (UW) distances in the tests below.

Test the hypothesis that microbiome is different between host species

```{r permanova-sp}
# With Wolbachia
HostSpecies <- sample_data(Wolb_rr_up_phytree)$Species
Year_all <- sample_data(Wolb_rr_up_phytree)$Year
Season_all <- sample_data(Wolb_rr_up_phytree)$Season
metadata <- data.frame(sample_data(Wolb_rr_up_phytree))
adonis2(formula = Wolb_unifrac_distUW ~ Species + Season_all, data = metadata, strata = Year_all, permutations = 999)

# Without Wolbachia
HostSpecies <- sample_data(NoWolb_rr_up_phytree)$Species
Year_all <- sample_data(NoWolb_rr_up_phytree)$Year
Season_all <- sample_data(NoWolb_rr_up_phytree)$Season
metadata <- data.frame(sample_data(NoWolb_rr_up_phytree))
adonis2(formula = NoWolb_unifrac_distUW ~ Species + Season_all, data = metadata, strata = Year_all, permutations = 999)
```

The microbiome of _Allograpta_ and _Eupeodes_ is different regardless of whether _Wolbachia_ is present or absent! It is worth noting that _Wolbachia_ itself is represented by different ASVs in each host species.


Now, to test the second hypothesis. To recap, we hypothesized:

"(2) That microbial communities are different between North and South populations for *non-migratory* hoverflies ( _Allograpta_ ), but more similar between North and South populations for *migratory* hoverflies ( _Eupeodes_ )"

First, we need to split datasets into species-specific data.

```{r split-data-in-species}
#Wolb
Allograpta_wolb <- Wolb_rr_up_phytree %>% subset_samples(Species == "Allograpta obliqua")
Eupeodes_wolb <- Wolb_rr_up_phytree %>% subset_samples(Species == "Eupeodes americanus")

#NoWolb
Allograpta_Nowolb <- NoWolb_rr_up_phytree %>% subset_samples(Species == "Allograpta obliqua")
Eupeodes_Nowolb <- NoWolb_rr_up_phytree %>% subset_samples(Species == "Eupeodes americanus")
```

Let's check if Bray-Curtis distances can be used to test geography, while respecting the assumption of homogeneity of variance for Permanova tests

```{r species-data-betadisper}
#Wolb
## Allograpta
Allograpta_wolbNS <- Allograpta_wolb %>% subset_samples(S_W_N != "West") # <- exclude a few "West" samples, so we can compare only North and South samples
otu_table <- t(as(otu_table(Allograpta_wolbNS), "matrix"))
AWolb_bray_dist <- vegdist(otu_table, method = "bray")
metadata_AWolb <- data.frame(sample_data(Allograpta_wolbNS))
disper_NS_AWolb <- betadisper(AWolb_bray_dist, metadata_AWolb$S_W_N)
anova(disper_NS_AWolb) # p-value < 0.0016; data is over dispersed
## Eupeodes
otu_table <- t(as(otu_table(Eupeodes_wolb), "matrix"))
EaWolb_bray_dist <- vegdist(otu_table, method = "bray")
metadata_EaWolb <- data.frame(sample_data(Eupeodes_wolb))
disper_NS_EaWolb <- betadisper(EaWolb_bray_dist, metadata_EaWolb$S_W_N)
anova(disper_NS_EaWolb) # p-value = 0.8578; data is not over dispersed

#NoWolb
## Allograpta
Allograpta_NowolbNS <- Allograpta_Nowolb %>% subset_samples(S_W_N != "West") # <- exclude a few "West" samples, so we can compare only North and South samples
otu_table <- t(as(otu_table(Allograpta_NowolbNS), "matrix"))
ANoWolb_bray_dist <- vegdist(otu_table, method = "bray")
metadata_ANoWolb <- data.frame(sample_data(Allograpta_NowolbNS))
disper_NS_ANoWolb <- betadisper(ANoWolb_bray_dist, metadata_ANoWolb$S_W_N)
anova(disper_NS_ANoWolb) # p-value = 0.1301; data is not over dispersed
## Eupeodes
otu_table <- t(as(otu_table(Eupeodes_Nowolb), "matrix"))
EaNoWolb_bray_dist <- vegdist(otu_table, method = "bray")
metadata_EaNoWolb <- data.frame(sample_data(Eupeodes_Nowolb))
disper_NS_EaNoWolb <- betadisper(EaNoWolb_bray_dist, metadata_EaNoWolb$S_W_N)
anova(disper_NS_EaNoWolb) # p-value = 0.4914; data is not over dispersed
```









```{r combine-bac-by-family, eval=FALSE}
pseq_alpha <-aggregate_taxa(Wolb_rr, level = "Genus")
tab <-microbiome::diversity(pseq_alpha, index = "Shannon")
```

Extract and prepare dataframe for plotting alpha diversity as a violin plot
Note: when plotting, the OnDemand rendering of Rstudio does not show some markdown text formatting, such as italicized or bold words
```{r plot-violin-plots}
pseq_alpha_meta <- microbiome::meta(pseq_alpha)
pseq_alpha_meta$Shannon <- tab$shannon

pseq_alpha_plot <- ggviolin(pseq_alpha_meta %>% mutate(Species = factor(Species, levels = c("Allograpta obliqua", "Eupeodes americanus", "Eupeodes pomus"))), x="Species", y="Shannon", add="boxplot", fill="Species")+scale_fill_manual(values = c("Allograpta obliqua" = "#FB9A99", "Eupeodes americanus" = "#E31A1C", "Eupeodes pomus" = "#FDBF6F"))

p1 <- pseq_alpha_plot+ labs(
  x = "Species",
  y = "Shannon index",
  title = "Microbiome Alpha Diversity in Hoverflies"
) +
  theme(
    axis.title.x = element_markdown(),
    axis.title.y = element_markdown()
  )
ggpar(p1, ylim = c(0, 3))
```















