---
title: "The Hoverfly microbiome: 16S data analysis"
author: "Pedro Rodrigues"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 3
    fig_caption: yes

fontsize: 14pt
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE)
```

```{r init, warning=FALSE, message=FALSE}
setwd("/Volumes/Extreme Pro/Syrphids16S/results_Syr16S/dada2")
set.seed("12345678")
library(dada2);packageVersion("dada2")
library(Biostrings); packageVersion("Biostrings")
library(rmarkdown)
library(tidyverse)
#BiocManager::install("BiocStyle")
```

```{r more preparation, warning=FALSE, message=FALSE}
path.out <- "Figures/"
path.rds <- "RDS/"
theme_set(theme_bw())
```

Load Data (quality-trimmed and denoised using DADA2 and the nf-core/ampliseq pipeline)
```{r phyloseq-load-improved-data, eval=FALSE}
p_data.chimera_free <- readRDS("DADA2_table.rds")
```

```{r assign taxonomy, eval=FALSE}
# Assign taxonomy
tax <- assignTaxonomy(p_data.chimera_free, "./rdp_train_set_18.fa.gz", minBoot = 80, tryRC = TRUE, multithread=TRUE) # Slowest part
head(unname(tax))
tax <- addSpecies(tax, "./rdp_species_assignment_18.fa.gz")
```

```{r save-progrss-taxonomy, eval=FALSE}
# Write to disk
#saveRDS(st.all_chimera_free, "RDS/PBB_all_runs_2.rds")
saveRDS(tax, "RDS/tax_hoverflies.rds")
saveRDS(p_data.chimera_free, "RDS/DADA2_table.rds")
```

> NOTE: taxonomy will be improved later, using BLAST (nt database), but only after the ASV table is further cleaned up in Phyloseq

## Move analysis over to Phyloseq

### Load all runs RDS files (ASV table file and taxonomy file)
#### Also load metadata file, where sample names are row names
```{r phyloseq-load-data, eval=FALSE}
library(tidyverse)
p_data.chimera_free <- readRDS("DADA2_table.rds")
tax <- readRDS("RDS/tax_hoverflies.rds")
metadata <- read.table("../input/2122metada.tsv", sep = "\t", header = TRUE)
samples.out <- data.frame(old_ID = rownames(p_data.chimera_free))
samples_out <- samples.out %>% separate_wider_delim(old_ID,delim ="_1.f", names = c("ID", "suffix"), cols_remove = FALSE)
metadata_merge <- left_join(metadata, samples_out, by = "ID")
row.names(metadata_merge) <- metadata_merge[,13]
#add isotope data to metadata
metadata_isotope <- read_csv("../input/metadata_isotope.csv")
metadata_merge_2 <- dplyr::left_join(metadata_merge, metadata_isotope, by="ID")
row.names(metadata_merge_2) <- metadata_merge_2[,13]
```

```{r backup-data, eval=FALSE}
write_tsv(metadata, "../input/metadataJan2025.tsv")
```
